File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int d = 0, curr = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        diameter(root);
        return d;
    }
    
    int diameter(TreeNode* root) {
        if(!root) return 0;
        int lh = diameter(root->left);
        int rh = diameter(root->right);
        curr = lh + rh;
        d = max(d, curr);
        return max(lh, rh) + 1;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int d = 0;
    int curr = 0;
    public int diam(TreeNode root) {
        if(root == null) return 0;
        int l = diam(root.left); int r = diam(root.right);
        curr = l + r;
        d = Math.max(d, curr);
        return Math.max(l, r) + 1;
    }
    public int diameterOfBinaryTree(TreeNode root) {
        diam(root);
        return d;
    }
}